llama.h

```diff
@@ -0,0 +1,139 @@
+#ifndef LLAMA_H
+#define LLAMA_H
+
+#include <stddef.h>
+#include <stdint.h>
+#include <stdbool.h>
+
+#ifdef LLAMA_SHARED
+#    ifdef _WIN32
+#        ifdef LLAMA_BUILD
+#            define LLAMA_API __declspec(dllexport)
+#        else
+#            define LLAMA_API __declspec(dllimport)
+#        endif
+#    else
+#        define LLAMA_API __attribute__ ((visibility ("default")))
+#    endif
+#else
+#    define LLAMA_API
+#endif
+
+#define LLAMA_FILE_VERSION 1
+#define LLAMA_FILE_MAGIC 0x67676d66 // 'ggmf' in hex
+#define LLAMA_FILE_MAGIC_UNVERSIONED 0x67676d6c // pre-versioned files
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+    //
+    // C interface
+    //
+    // TODO: show sample usage
+    //
+
+    struct llama_context;
```

Comment:
It would be really cool if the memory in the llama context and the actual model parameters were in separate structs, so you could initialize a model from a file, then create multiple contexts from it and inference them in parallel. this would be useful if you want to have multiple prompts going at the same time so they can all share the model weights, but have their own k/v memory

Reply:
I think it would also be useful if the thread safety of each function was mentioned. For example, can I call `llama_eval()` and `llama_tokenize()` concurrently?

Reply:
Yes, this will be added via `llama_state` similar to the way it is implemented in `whisper.cpp`.
I didn't include it in the first pass of the implementation in order to deal with a smaller change. Will probably add this on Thursday with a second pass. It will not change the existing API - it will simply extend it with new per-state calls.